services/MerchStoreService.ts

```diff
@@ -291,6 +291,68 @@ export default class MerchStoreService {
       await user.reload();
       const merchItemOptionRepository = Repositories.merchStoreItemOption(txn);
       const itemOptions = await merchItemOptionRepository.batchFindByUuid(originalOrder.map((oi) => oi.option));
+
+      const totalCost = await this.verifyOrder(originalOrder, user);
+
+      const merchOrderRepository = Repositories.merchOrder(txn);
+
+      // if all checks pass, the order is placed
+      const createdOrder = await merchOrderRepository.createMerchOrder(OrderModel.create({
+        user,
+        totalCost,
+        items: flatten(originalOrder.map((optionAndQuantity) => {
+          const option = itemOptions.get(optionAndQuantity.option);
+          const quantityRequested = optionAndQuantity.quantity;
+          return Array(quantityRequested).fill(OrderItemModel.create({
+            option,
+            salePriceAtPurchase: option.getPrice(),
+            discountPercentageAtPurchase: option.discountPercentage,
+          }));
+        })),
+      }));
+
+      const activityRepository = Repositories.activity(txn);
+      await activityRepository.logActivity({
+        user,
+        type: ActivityType.ORDER_MERCHANDISE,
+        description: `Order ${createdOrder.uuid}`,
+      });
+
+      await Promise.all(originalOrder.map(async (optionAndQuantity) => {
+        const option = itemOptions.get(optionAndQuantity.option);
+        const updatedQuantity = option.quantity - optionAndQuantity.quantity;
+        return merchItemOptionRepository.upsertMerchItemOption(option, { quantity: updatedQuantity });
+      }));
+
+      const userRepository = Repositories.user(txn);
+      userRepository.upsertUser(user, { credits: user.credits - totalCost });
+
+      return [createdOrder, itemOptions];
+    });
+
+    const orderConfirmation = {
+      items: originalOrder.map((oi) => {
+        const option = merchItemOptions.get(oi.option);
+        const { item } = option;
+        return {
+          ...item,
+          quantityRequested: oi.quantity,
+          salePrice: option.getPrice(),
+          total: oi.quantity * option.getPrice(),
+        };
+      }),
+      totalCost: order.totalCost,
+    };
+    this.emailService.sendOrderConfirmation(user.email, user.firstName, orderConfirmation);
+
+    return order.getPublicOrder();
+  }
+
+  public async verifyOrder(originalOrder: MerchItemOptionAndQuantity[], user: UserModel): Promise<number> {
+    const totalOrderCost = await this.transactions.readWrite(async (txn) => {
+      await user.reload();
+      const merchItemOptionRepository = Repositories.merchStoreItemOption(txn);
+      const itemOptions = await merchItemOptionRepository.batchFindByUuid(originalOrder.map((oi) => oi.option));
```

Comment:
nit: you can just do `return this.transactions.readWrite(async (txn))...` instead of awaiting the call and returning the same value returned by the async call, since we're returning a `Promise` anyway.